### PR TITLE
Added WebPageDigester class. Slight modifications to WebPage.

### PR DESCRIPTION
**Pull request by lgfrbcsgo**
- Added `WebPageDigester` class for adding occurrences to `WebPage` objects. It will filter the content of a passed `WebPage` for search term occurrences with the set filter and optional pre-filter. The result will be added to the passed `WebPage` object. Occurrences can be merged (with optional parameter `mergeOccurrences`) or overwritten.

**USAGE:**
```
import {WebPageDigester} from "../webpage-digester";
import {PrefixTree} from "../filters/prefix-tree";
import {WebPage} from "../utils/webpage";

let terms = ["some", "terms"];
let digester = new WebPageDigester(terms).setFilter(PrefixTree);
digester.digest(new WebPage());
```
Allows changing and removing filters and method chaining:
```
let digester = new WebPageDigester(terms)
    .setFilter(PrefixTree)
    .setPreFilter(BloomFilter)
    .removePreFilter()
    .setPreFilter(PrefixTree)
    .setFilter(NaiveFilter);
```

- Reworked `.getTLD()` in `WebPage`, added `occurrence` property and made constructor parameter optional

- Added UNIT tests for `WebPageDigester` and `WebPage`

- Renamed `web-page.ts` to `webpage.ts`

### CODE REVIEW

File: src/webpage-digester.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+            let matches = this.preFilterInstance.getMatches(pageContent);
+
+            // Stop here if there aren't any matches.
+            if (matches.size === 0) {
+                return webPage;
+            }
+
+            // create new IndexFilter instance from matched searchTerms
+            this.mainFilterInstance = new this.mainFilter();
+            this.mainFilterInstance.addSearchTerms([...matches]); // [... Set] operator converts Set to Array
```
Comment (nyxathid) [inline src/webpage-digester.ts]: If we have a prefilter, do we create a new mainFilterInstance for EACH new WebPage that is digested? That might actually be slower than not using a prefilter at all.

File: package.json [hunk 2]
```diff
@@ -22,7 +22,7 @@
     "chai": "^3.5.0"
   },
   "scripts": {
-    "start": "node app.js",
+    "start": "node out/app.js",
```
Comment (nyxathid) [inline package.json]: I'm wondering if we should add a `tsc` call before that. Using WebStorm/IntelliJ our `.ts` files are compiled automatically on save, but on our Azure VMs we will have to run `tsc` manually.

File: src/webpage-digester.ts (hunk 1 above)
Comment (lgfrbcsgo) [inline src/webpage-digester.ts]: Yes, but only if we've found a match with the pre-filter. Intended use case is for example `PrefixTree` as pre-filter and `NaiveFilter` as main filter. Reason being that `PrefixTree` is a lot faster than `NaiveFilter`, but accepts pre and suffixes on search terms which might not be desired.

File: package.json (hunk 2 above)
Comment (lgfrbcsgo) [inline package.json]: Good point!

File: src/webpage-digester.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+     * This is very useful if your main filter is slow
+     * @param filterConstructor                             Class of the pre-filter
+     */
+    public setPreFilter<T extends Filter> (filterConstructor : new () => T) : WebPageDigester{
+        this.preFilterInstance = new filterConstructor();
+        this.preFilterInstance.addSearchTerms(this.searchTerms);
+        return this;
+    }
+
+    public removePreFilter() : WebPageDigester {
```
Comment (nyxathid) [inline src/webpage-digester.ts]: we could also name it `removeFilters` or `removeAllFilters`

Comment (lgfrbcsgo) [inline src/webpage-digester.ts]: It only removes the pre-filter. You currently can only have a singe pre-filter. I couldn't think of a use case for multiple pre-filters.

File: src/webpage-digester.ts (hunk 1 above)
Comment (nyxathid) [inline src/webpage-digester.ts]: OK, this totally makes sense.
For this case we could also override the `addSearchTerms()` function in `NaiveFilter`: it doesn't call `addSearchTerm` for each term but just saves the whole array (`searchTerms`)

File: src/webpage-digester.ts (hunk 3 above)
Comment (nyxathid) [inline src/webpage-digester.ts]: Ah, I see 👍

**Comment by nyxathid** [comment]
Very good job! :D

File: src/webpage-digester.ts (hunk 1 above)
Comment (lgfrbcsgo) [inline src/webpage-digester.ts]: The problem with `addSearchTerms()` is that it adds terms, not overwrites the current ones. The best way to do this is to standardize the constructor function. Was a bit lazy there and just used `addSearchTerms()` 😄

Comment (lgfrbcsgo) [inline src/webpage-digester.ts]: Well, sorry I can't seem to get that to work. I managed to replace the constructor function by a protected `.init()` function which gets called by the super constructor. I works all nicely until you try to pass an argument to the constructor of a wildcard class....

**Pull request merged**